universal-hash/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+//! as their construction has a number of desirable properties such as
+//! pairwise independence as well as amenability to efficient implementations,
+//! particularly when implemented using SIMD instructions.
+//!
+//! When combined with a cipher, such as in Galois/Counter Mode or the
+//! Salsa20 family AEAD constructions, they can provide the core functionality
+//! for a Message Authentication Code (MAC).
+//!
+//! [Universal Hash Functions]: https://en.wikipedia.org/wiki/Universal_hashing
+
+#![no_std]
+#![doc(html_logo_url = "https://raw.githubusercontent.com/RustCrypto/meta/master/logo_small.png")]
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+use generic_array::{ArrayLength, GenericArray};
+use generic_array::typenum::Unsigned;
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash: Clone {
+    /// Size of a block (e.g. field element) this universal hash operates on
+    type BlockSize: ArrayLength<u8>;
+
+    /// Instantiate a universal hash function with the given key
+    fn new(key: &GenericArray<u8, Self::BlockSize>) -> Self;
+
+    /// Input a block into the universal hash function
+    fn update_block(&mut self, block: &GenericArray<u8, Self::BlockSize>);
+
+    /// Input data into the universal hash function. If the length of the
+    /// data is not a multiple of the block size, the remaining data is
+    /// padded with zeroes up to the `BlockSize`.
+    ///
+    /// This approach is frequently used by AEAD modes which use
+    /// Message Authentication Codes (MACs) based on universal hashing.
+    fn update_padded(&mut self, data: &[u8]) {
+        let mut chunks = data.chunks_exact(Self::BlockSize::to_usize());
+
+        for chunk in &mut chunks {
+            self.update_block(GenericArray::from_slice(chunk));
+        }
+
+        let rem = chunks.remainder();
+
+        if !rem.is_empty() {
+            let mut padded_block = GenericArray::default();
+            padded_block[..rem.len()].copy_from_slice(rem);
+            self.update_block(&padded_block);
+        }
+    }
+
+    /// Reset `UniversalHash` instance.
+    fn reset(&mut self);
+
+    /// Obtain the [`Output`] of a `UniversalHash` function and consume it.
+    fn result(self) -> Output<Self::BlockSize>;
+
+    /// Obtain the [`Output`] of a `UniversalHash` computation and reset it back
+    /// to its initial state.
+    fn result_reset(&mut self) -> Output<Self::BlockSize> {
+        let res = self.clone().result();
+        self.reset();
+        res
+    }
+
+    /// Verify the `UniversalHash` of the processed input matches a given [`Output`].
+    /// This is useful when constructing Message Authentication Codes (MACs)
+    /// from universal hash functions.
+    fn verify(self, other: impl Into<Output<Self::BlockSize>>) -> Result<(), Error> {
```

Comment:
Why do we need this generality here? I think using something like this would be good enough:
```rust
fn verify(self, other: &GenericArray<u8, Self::BlockSize>) -> Result<(), Error> {
    if self.result() == other.into() {
        Ok(())
    } else {
        Err(MacError)
    }
}
```

Reply:
Ok I'll get rid of it for now